controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp

```diff
@@ -102,28 +102,24 @@ void dynamic_channel_selection_task::work()
         break;
     }
     case eState::IDLE: {
-        if (m_is_single_scan_pending) {
+        if (m_is_single_scan_pending && start_scan()) {
```

Comment:
IMO It's better to call this function right before moving to the `TRIGGER_SCAN` state.
Something like this:
```c++
auto now = std::chrono::steady_clock::now();
if (m_is_single_scan_pending) {
    ...
} else if (database.get_channel_scan_is_enabled(m_radio_mac) {
    ...
} else {
    break;
}

if (start_scan()) {
    m_last_scan_try_timestamp = now;
    fsm_move_state(eState::TRIGGER_SCAN);
}
```

In this case you actually start the scan (`start_scan()`) only when all conditions are met and no other task is currently scanning.

Reply:
but the solution you suggest updates the `m_last_scan_try_timestamp` for a single scan as well.

Reply:
Also true, but can be solved by adding another state :)
`IDLE` -> `CAN_SCAN` -> `TRIGGER_SCAN` -> ... -> `FINISH`

The FSM will reach the `CAN_SCAN` state only when it's ready to scan.
It'll wait in this state until it can actually trigger a scan, and only then move to `TRIGGER_SCAN`.
`FINISH` will always do all the cleanups (including removing the id of the scanning task).

Reply:
it is more of `SHOULD_SCAN` rather than `CAN_SCAN` since the `start_scan()` itself is the condition of can-scan

Reply:
I added a PRE_SCAN state and FINISH state - please re-review